compiler/optimizations/optimizeOnClauses.cpp

```diff
@@ -411,30 +446,77 @@ optimizeOnClauses(void) {
   if (fNoOptimizeOnClauses)
     return;
 
+  // If we're not using locks, the extern functions in
+  // atomics are local and safe for fast on.
+  if (0 != strcmp(CHPL_ATOMICS, "locks")) {
+    forv_Vec(ModuleSymbol, module, gModuleSymbols) {
+      if( module->hasFlag(FLAG_ATOMIC_MODULE) ) {
+        Vec<FnSymbol*> moduleFunctions = module->getTopLevelFunctions(true);
+        forv_Vec(FnSymbol, fn, moduleFunctions) {
+          if( fn->hasFlag(FLAG_EXTERN) ) {
+            fn->addFlag(FLAG_FAST_ON_SAFE_EXTERN);
+            fn->addFlag(FLAG_LOCAL_FN);
+          }
+        }
+      }
+    }
+  }
```

Comment:
Anyway, the only thing we are worried about here is if the intrinsic might make a *system call*. It could be implemented with a spin lock. Even if implemented with a lock, I think it would be unreasonable to try to handle contention in fetch_and_add by swapping to a different thread, for example.

Anyway, if we run into a platform that does this, we can just turn off atomics=intrinsics for it.